### ISSUE TITLE
500 when creating new product after adding attribute via API and assigning to attribute set via UI

### ISSUE THREAD

**Issue by vincentteyssier** [magento/magento2#26827]
### Preconditions (*)
Magento 2.3.4
PHP 7.3

### Steps to reproduce (*)
1. Create new attribute using API

```
// End point: POST "/rest/default/V1/products/attributes";

// prepare the payload
$data = array(
  "attribute" => array(
    "is_wysiwyg_enabled" => false,
    "is_html_allowed_on_front" => true,
    "used_for_sort_by" => true,
    "is_filterable" => true,
    "is_filterable_in_search" => true,
    "is_used_in_grid" => true,
    "is_visible_in_grid" => true,
    "is_filterable_in_grid" => true,
    "position" => 0,
    "is_searchable" => true,
    "is_visible_in_advanced_search" => true,
    "is_comparable" => true,
    "is_used_for_promo_rules" => false,
    "is_visible_on_front" => true,
    "used_in_product_listing" => true,
    "is_visible" => true,
    "attribute_code" => "test_api",
    "frontend_input" => "swatch_text",
    "entity_type_id" => "4",
    "is_required" => true,
    "is_user_defined" => true,
    "default_frontend_label" => "test_api2",
    "frontend_labels" => [
      array(
        "store_id" => 0,
        "label" => "test_api3"
      )
    ],
    "backend_type" => "int",
    "is_unique" => false,
    "is_required" => false
  )
);

$payload = json_encode($data);

// make the call
$ch = curl_init();
curl_setopt($ch, CURLOPT_USERAGENT, "Mozilla/5.0 (Windows; U; Windows NT 5.1; rv:1.7.3) Gecko/20041001 Firefox/0.10.1");
curl_setopt($ch, CURLOPT_HTTPHEADER, array('Content-Type: application/json' , $auth ));
curl_setopt($ch, CURLOPT_POST, 1);
curl_setopt($ch, CURLOPT_POSTFIELDS, $payload);
curl_setopt($ch, CURLOPT_URL, $url);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch, CURLOPT_AUTOREFERER, true);
curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
$result = curl_exec($ch);
curl_close($ch);
```

2. Log in to Admin panel, go to Stores/Attributes/Attributes Set
3. Add the newly created attrbute to the default attribute set
4. Go to product page and click ADD PRODUCT

### Expected result (*)
Product creation page should display normally

### Actual result (*)
Error 500
Nginx logs below:

> 2020/02/12 04:45:19 [error] 1023#1023: *46116 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught TypeError: Argument 2 passed to Magento\Framework\View\Element\UiComponentFactory::argumentsResolver() must be of the type array, null given, called in /opt/magento/public_html/vendor/magento/framework/View/Element/UiComponentFactory.php on line 222 and defined in /opt/magento/public_html/vendor/magento/framework/View/Element/UiComponentFactory.php:174
> Stack trace:
> #0 /www/vendor/magento/framework/View/Element/UiComponentFactory.php(222): Magento\Framework\View\Element\UiComponentFactory->**argumentsResolver('test_api', NULL)**
> #1 /www/vendor/magento/module-ui/Component/Form/Field.php(85): Magento\Framework\View\Element\UiComponentFactory->**create('test_api', 'swatch_text', Array)**
> #2 /www/vendor/magento/framework/View/Layout/Generator/UiComponent.php(164): Magento\Ui\Component\Form\Field->prepare()
> #3 /www/vendor/magento/framework/View/Layout/Generator/UiComponent.php(161): Magento\Fr" while reading response header from upstream, client: 127.0.0.1, server: example.com, request: "GET /admin_xxx/catalog/product/new/set/4/type/simple/key/zzzzzzzzzzz/ HTTP/1.1", upstream: "fastcgi://unix:/var/run/php/php7.3-fpm-magento.sock:", host: "example.com", referrer: "https://example.com/admin_xxx/catalog/product/index/key/yyyyyyyyyy/"
>

**Comment by vincentteyssier** [comment]
Additional comment: if I go to the test attribute page and save it without modifying anything the 500 does not occur anymore

**Comment by vincentteyssier** [comment]
I digged a bit into the code and can see that what gives NULL in argumentsResolver('test_api', NULL) comes from $rawComponentData = $this->definitionData->get($name) and $name is 'swatch_text' as we can see in the create instruction.
However I can't find anywhere allowed values for this. I guess this is the frontend_input parameter in the API call. I've seen a few examples with this value so I assumed it was allowed. Can anyone help me find the allowed values?

**Comment by vincentteyssier** [comment]
Interestingly the indexer is failing to reindex tables:

**SELECT `e`.`entity_id`, `e`.`type_id`, `e`.`sku`, **`e`.`test_api`** FROM `catalog_product_entity` AS `e`

INNER JOIN `catalog_product_website` AS `website` ON website.product_id = e.entity_id AND website.website_id = 1

INNER JOIN `catalog_product_entity_int` AS `visibility_default` ON visibility_default.entity_id= e.entity_id AND visibility_default.attribute_id = '99' AND visibility_default.store_id = 0

LEFT JOIN `catalog_product_entity_int` AS `visibility_store` ON visibility_store.entity_id= e.entity_id AND visibility_store.attribute_id = '99' AND visibility_store.store_id = '1'

INNER JOIN `catalog_product_entity_int` AS `status_default` ON status_default.entity_id= e.entity_id AND status_default.attribute_id = '97' AND status_default.store_id = 0

LEFT JOIN `catalog_product_entity_int` AS `status_store` ON status_store.entity_id= e.entity_id AND status_store.attribute_id = '97' AND status_store.store_id = '1' WHERE (IF(visibility_store.value_id > 0, visibility_store.value, visibility_default.value) IN (3, 2, 4)) AND (IF(status_store.value_id > 0, status_store.value, status_default.value) IN (1)) AND (e.entity_id > 0) AND (e.entity_id < 501) ORDER BY `e`.`entity_id` ASC

LIMIT 500**

It is expecting the table catalog product entity to have a test_api column.... this is surprising

**Comment by vincentteyssier** [comment]
Also notice the reindexer is failing, even when I set searchable fields to false:

```
Catalog Search indexer process unknown error:
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'e.test_api' in 'field list', query was: SELECT `e`.`entity_id`, `e`.`type_id`, `e`.`sku`, `e`.`test_api` FROM `catalog_product_entity` AS `e`
 INNER JOIN `catalog_product_website` AS `website` ON website.product_id = e.entity_id AND website.website_id = 1
 INNER JOIN `catalog_product_entity_int` AS `visibility_default` ON visibility_default.entity_id= e.entity_id AND visibility_default.attribute_id = '99' AND visibility_default.store_id = 0
 LEFT JOIN `catalog_product_entity_int` AS `visibility_store` ON visibility_store.entity_id= e.entity_id AND visibility_store.attribute_id = '99' AND visibility_store.store_id = '1'
 INNER JOIN `catalog_product_entity_int` AS `status_default` ON status_default.entity_id= e.entity_id AND status_default.attribute_id = '97' AND status_default.store_id = 0
 LEFT JOIN `catalog_product_entity_int` AS `status_store` ON status_store.entity_id= e.entity_id AND status_store.attribute_id = '97' AND status_store.store_id = '1' WHERE (IF(visibility_store.value_id > 0, visibility_store.value, visibility_default.value) IN (3, 2, 4)) AND (IF(status_store.value_id > 0, status_store.value, status_default.value) IN (1)) AND (e.entity_id > 0) AND (e.entity_id < 501) ORDER BY `e`.`entity_id` ASC
 LIMIT 500
```

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Delta 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-31474` were created

**Issue Available**: @engcom-Delta, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by vincent-le89** [comment]
Hi guys,

This issue happens when we save a product attribute with "frontend_input" => "swatch_text" via API. Other front_input types like text,... haven't happened.
We need someone checkout logic of repository: function save(app/code/Magento/Catalog/Model/Product/Attribute/Repository.php).
@vincentteyssier After you create a product attribute via API, you can go to that product attribute in admin and save it again. This error will be resolved.
I think this issue will take a lot of time to resolve it.
Cheers,

**Comment by vincentteyssier** [comment]
Hi and thanks for looking into this. The workaround is a bit tedious since I might have to do that for hu ndreds of products, but not much of a choice :)
I reported another issue with the attributes option API when attribute is swatch_text.... not sure this helps, but there seems to be a few issues related to swatch text. 
Thanks a lot and good luck!

**Comment by magento-engcom-team** [comment]
Hi @vincentteyssier. Thank you for your report. 
The issue has been fixed in magento/magento2#27191 by @sergiy-v in 2.4-develop branch
Related commit(s):
 - [acfc589b812278750e6545edc9f0e7b71a80cea3](https://github.com/magento/magento2/commit/acfc589b812278750e6545edc9f0e7b71a80cea3)
 - [fcf734fc689da3bca3fd8c1c0800204540de1969](https://github.com/magento/magento2/commit/fcf734fc689da3bca3fd8c1c0800204540de1969)

The fix will be available with the upcoming 2.4.0 release.
